src/Bramus/Router/Router.php

```diff
@@ -288,13 +288,11 @@ public function run($callback = null)
         }
 
         // If no route was handled, trigger the 404 (if any)
-        if ($numHandled === 0) {
+        if ($numHandled === 0 && isset($this->afterRoutes[$this->requestedMethod])) {
             $this->trigger404($this->afterRoutes[$this->requestedMethod]);
         } // If a route was handled, perform the finish callback (if any)
-        else {
-            if ($callback && is_callable($callback)) {
-                $callback();
-            }
+        elseif ($callback && is_callable($callback)) {
```

Comment:
By introducing the `elseif`, it also introduces the unwanted side-effect of running the callback when there is no handler and no afterRoute … To not change the existing behavior, the callback should only be run when there was a handler.

Reply:
The solution lies at line 291: don’t `&&` the extra condition, but nest it. Then this `elseif` could remain in place.